File: xaiop-sdk/go/xaiop/label_escape.go
package xaiop

// LabelEscapeIntroducer is U+001F for symbol-key mode.
const LabelEscapeIntroducer = "\u001f"

// KeyNeedsSymbolEscape reports whether a JSON key must be escaped on the wire
// when symbolKeys is enabled (operator-headed or introducer-headed keys).
func KeyNeedsSymbolEscape(key string) bool {
	if key == "" {
		return false
	}
	c := key[0]
	return c == 0x1F || c == '#' || c == '@' || c == '>' || c == '<' || c == '=' || c == '!' || c == '&' || c == '?'
}

// EncodeWireLabel escapes a key for the wire when symbolKeys is on.
func EncodeWireLabel(key string, symbolKeys bool) string {
	if symbolKeys && KeyNeedsSymbolEscape(key) {
		return LabelEscapeIntroducer + key
	}
	return key
}

// DecodeWireLabel strips the U+001F introducer when symbolKeys is on.
func DecodeWireLabel(wireLabel string, symbolKeys bool) string {
	if symbolKeys && len(wireLabel) > 0 && wireLabel[0] == 0x1F {
		return wireLabel[1:]
	}
	return wireLabel
}

File: xaiop-sdk/python/src/xaiop/label_escape.py
"""Symbol-key mode: U+001F label escape for operator-headed JSON keys."""

from __future__ import annotations

LABEL_ESCAPE_INTRODUCER = "\u001f"


def key_needs_symbol_escape(key: str) -> bool:
    if not isinstance(key, str) or not key:
        return False
    c = ord(key[0])
    return c in (0x1F, 0x23, 0x40, 0x3E, 0x3C, 0x3D, 0x21, 0x26, 0x3F)


def encode_wire_label(key: str, symbol_keys: bool) -> str:
    if symbol_keys and key_needs_symbol_escape(key):
        return LABEL_ESCAPE_INTRODUCER + key
    return key


def decode_wire_label(wire_label: str, symbol_keys: bool) -> str:
    if (
        symbol_keys
        and isinstance(wire_label, str)
        and wire_label
        and ord(wire_label[0]) == 0x1F
    ):
        return wire_label[1:]
    return wire_label
